Support Python Slice class in TorchScript

gmagogsfm opened the pull request:
Implements support for[ Python Slice class](https://docs.python.org/3/c-api/slice.html) (not slice expression, which is already supported)

Slice object can be used in any place that supports slice expression, including multi-dim tensor slicing.

Fixes #43511
Fixes #43125

gmagogsfm committed:
Implement Python slice

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3297,52 +3297,79 @@ struct to_ir {
     }
+    args.emplace_back(loc, "start", start);
 
-    args.emplace_back(loc, "begin", emitExpr(Expr(slice.startOr(0))));
```
Note to reviewer:

I changed name of this argument from "begin" to "start" to match the schema of `aten::slice`. I think it worked before because this argument is always used as positional argument. Let me know if there are other things to consider.

SplitInfinity commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3705,17 +3763,43 @@ struct to_ir {
+      if (const auto psv =
+              dynamic_cast<PythonSliceValue*>(subscript_sv.get())) {
+        // aten::slice.tensor needs an additional `dim` input.
+        if (sliceable->type()->isSubtypeOf(TensorType::get())) {
+          Value* dim = method.graph()->insertConstant(0, val_range);
+          Value* sliced = emitBuiltinCall(
+              val_range,
+              *method.graph(),
+              aten::slice,
+              {sliceable, dim, psv->start(), psv->stop(), psv->step()},
+              {});
+          return std::make_shared<SimpleValue>(sliced);
+        }
+
+        Value* sliced = emitBuiltinCall(
+            val_range,
+            *method.graph(),
+            aten::slice,
+            {sliceable, psv->start(), psv->stop(), psv->step()},
+            {});
+        return std::make_shared<SimpleValue>(sliced);
```
These branches look identical except for `dim`. Consider using a vector instead of array literals for the method args so you can reuse the `emitBuiltinCall` and `return`.

gmagogsfm replied:
Done.

SplitInfinity commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3705,17 +3763,43 @@ struct to_ir {
       AT_ASSERT(subscript_exprs.size() == 1);
+      Value* sliceable = sv->asValue(val_range, method);
+
+      auto subscript_sv = emitSugaredExpr(subscript_exprs[0], 1);
```
Consider adding a comment like the one above which explains that this code handles slice objects.

gmagogsfm replied:
Done.

SplitInfinity commented on torch/csrc/jit/frontend/sugared_value.h:
```diff
@@ -772,5 +772,29 @@ struct TORCH_API SugaredEnumClass : public SugaredValue {
   EnumTypePtr enum_type_;
 };
 
+struct TORCH_API PythonSliceValue : public SugaredValue {
```
Maybe `SliceValue` because this is a regular `SugaredValue`?

gmagogsfm replied:
This is actually something I want to get advice on. It is a vanilla `SugaredValue` only because `ir_emitter` needs it and `ir_emitter` can't depend on `python_sugared_value`. In reality, `PythonSliceValue` does represent a python object, thus the `python` prefix. Let me know what you think.

SplitInfinity replied:
Hm, there is a `ClassValue` class that also extends `SugaredValue` that represents Python classes. So I think `SliceValue` would be better here.

gmagogsfm replied:
Done.

SplitInfinity reviewed (commented):
What is the motivation for this work? What can slice objects support that slice expressions do not already support?

SplitInfinity commented on test/jit/test_slice.py:
```diff
@@ -0,0 +1,101 @@
+                       "instead.")
+
+# Tests that Python slice class is supported in TorchScript
+class TestSlice(JitTestCase):
```
Consider adding a test in which `stop` is greater than the length of the iterable to check that out of bounds indices are clipped correctly.

gmagogsfm replied:
Done.

gmagogsfm commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -785,6 +786,56 @@ std::shared_ptr<SugaredValue> createSimpleEnumValue(
+    throw ErrorReport(loc) << "Slice does not accept any keyword arguments";
+  }
+
+  static const int64_t default_start = 0;
```
Done.

gmagogsfm commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -785,6 +786,56 @@ std::shared_ptr<SugaredValue> createSimpleEnumValue(
+  }
+
+  static const int64_t default_start = 0;
+  static const int64_t default_stop = LLONG_MAX;
```
Done.

gmagogsfm commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -785,6 +786,56 @@ std::shared_ptr<SugaredValue> createSimpleEnumValue(
+
+  static const int64_t default_start = 0;
+  static const int64_t default_stop = LLONG_MAX;
+  static const int64_t default_step = 1;
```
Done.

gmagogsfm commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -785,6 +786,56 @@ std::shared_ptr<SugaredValue> createSimpleEnumValue(
+  static const int64_t default_start = 0;
+  static const int64_t default_stop = LLONG_MAX;
+  static const int64_t default_step = 1;
+  Graph& graph = *(caller.graph());
```
Any suggestion on how to avoid it? Adding an assert here to ensure `graph` is not null? I though it is an invariant that `graph` is always not null when emitting IR. Otherwise, most of our code would segfault.

gmagogsfm commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -785,6 +786,56 @@ std::shared_ptr<SugaredValue> createSimpleEnumValue(
+  Graph& graph = *(caller.graph());
+
+  auto ValOr = [&](Value* given, int64_t default_val) {
+    if (!given || given->type()->isSubtypeOf(NoneType::get())) {
```
`isSubtypeOf` is a superset check of `equals`. Searching through code base it seems this is a common pattern to check subtype of None. As for the reason behind it, I guess it is just to be future proof.

gmagogsfm commented on torch/csrc/jit/python/python_sugared_value.h:
```diff
@@ -305,5 +305,21 @@ struct VISIBILITY_HIDDEN PythonExceptionValue : public ExceptionValue {
 };
 
+// Python Slice class.
+struct VISIBILITY_HIDDEN PythonSliceClass : public PythonValue {
```
Done.

gmagogsfm reviewed (commented):
As for motivation, this is to fix #43511

The only one case where a slice object is better than slice expression, that is to share slice expression among different objects. Otherwise they are equivalent. So motivation is to fix the issue and enable one more language feature to save users some work-around time.

SplitInfinity reviewed (approved):
Why did so few CI jobs run on this? Please try to get that extra signal before attempting to land this.

gmagogsfm commented:
> Why did so few CI jobs run on this? Please try to get that extra signal before attempting to land this.

Because I rebased this morning to circumvent a Phabricator patch issue. You happened to open this PR when I just rebased, so CIs are just beginning.
